### PR TITLE
Tweak: Adding styled-components infrastructure [ED-4561]

### PR DESCRIPTION
**Pull request by rotemee**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

**Review by matipojo — CHANGES_REQUESTED** [review]
🔥

File: packages/elementor-ui/components/utils.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/**
+ * @param styles - { base: { shared: '', variant: { h1: '', h2: '' } }, dark: { shared: '', variant: { h1: '', h2: '' } } }
```
Comment (kobizz) [inline packages/elementor-ui/components/utils.js:2]: Can we make the shared CSS to be rendered only once per component type?

File: packages/elementor-ui/components/utils.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		const themeVariant = styles[ key ];
+
+		// If key exist in the styles obj (dark, light etc.)
+		if ( themeVariant ) {
+			if ( 'string' === typeof themeVariant ) {
+				variantStyle += themeVariant;
+			} else {
+				variantStyle += themeVariant.shared || '';
+			}
+
```
Comment (kobizz) [inline packages/elementor-ui/components/utils.js:34]: Try making it an inner function the can be reused below

File: packages/elementor-ui/styles/_data/breakpoints/breakpoints.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (kobizz) [inline packages/elementor-ui/styles/_data/breakpoints/breakpoints.json:1]: Remove

File: packages/elementor-ui/styles/_maps/breakpoints/breakpoints.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import breakpoints from '../../_data/breakpoints/breakpoints.json';
```
Comment (kobizz) [inline packages/elementor-ui/styles/_maps/breakpoints/breakpoints.js:1]: Remove

**Review by kobizz — CHANGES_REQUESTED** [review]

File: packages/elementor-ui/components/text.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { forwardRef } from 'react';
```
[the remark this answers is not in the document]
Comment (rotemee) [reply packages/elementor-ui/components/text.js:1]: Done.

File: packages/elementor-ui/styles/_data/typography/typography.json [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  "text-xxs-font-size": "0.6875rem",
+  "font-size-micro": "0.625rem",
+
+  "body-font-size": "0.875rem",
+  "font-size-caption": "0.75rem",
+
+  "small-font-size": "80%",
+  "paragraph-margin-bottom": "",
+  "headings-margin-bottom": ""
+
```
[the remark this answers is not in the document]
Comment (rotemee) [reply packages/elementor-ui/styles/_data/typography/typography.json:48]: Done.

File: packages/elementor-ui/styles/_data/typography/typography.json [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  "line-height-flat": 1,
+  "line-height-base": 1.5,
+  "line-height-sm": 1.2,
+  "line-height-lg": 1.8,
+
+  "display-1-font-size": "1.85rem",
+  "display-2-font-size": "1.85rem",
+  "display-3-font-size": "1.85rem",
+  "display-4-font-size": "1.85rem",
```
[the remark this answers is not in the document]
Comment (rotemee) [reply packages/elementor-ui/styles/_data/typography/typography.json:25]: Fixed.

File: packages/elementor-ui/components/heading.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import { forwardRef } from 'react';
+import { Styled, withStyles } from 'e-components';
+import styles from 'e-styles/heading';
+
+const Heading = forwardRef( ( props, ref ) => (
+	<Styled { ...props } tag="h1" styles={ styles } ref={ ref }>{ props.children }</Styled>
```
[the remark this answers is not in the document]
Comment (rotemee) [reply packages/elementor-ui/components/heading.js:6]: Fixed.

File: packages/elementor-ui/components/styled.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+	// Removing props that are not related to the component styles object.
+	nonStyledProps.forEach( ( prop ) => delete styledProps[ prop ] );
+
+	const stylesData = { props: styledProps, config: themeContext.config },
+		componentCss = getStyle( props.styles, stylesData ),
+		stylesExtension = getStyle( props.extendStyles, stylesData ),
+		StyledComponent = styled[ props.tag ]`${ componentCss }${ stylesExtension }`;
+
+	return <StyledComponent { ...props } ref={ ref }>{ props.children }</StyledComponent>;
```
[the remark this answers is not in the document]
Comment (rotemee) [reply packages/elementor-ui/components/styled.js:18]: My mistake, basically we can keep it for some reason that someone will need inline styling.

File: packages/elementor-ui/components/styled.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import { useContext, forwardRef } from 'react';
+import styled, { ThemeContext } from 'styled-components';
+import { getStyle } from 'e-components/utils';
+
+const Styled = forwardRef( ( props, ref ) => {
+	const themeContext = useContext( ThemeContext ),
+		styledProps = { ...props },
+		nonStyledProps = [ 'children', 'className', 'component', 'tag', 'styles', 'extendStyles', 'extend' ];
```
[the remark this answers is not in the document]
Comment (rotemee) [reply packages/elementor-ui/components/styled.js:8]: Done.

File: packages/elementor-ui/styles/_data/colors/palette.json [hunk 11]
```diff
@@ -0,0 +1,122 @@
+{
+  "white": {
+	"hex": "#fff",
+	"rgb": "rgb(255, 255, 255)",
+	"hsl": "hsl(0, 0%, 100%)"
```
[the remark this answers is not in the document]
Comment (rotemee) [reply packages/elementor-ui/styles/_data/colors/palette.json:5]: Removed.

File: packages/elementor-ui/components/utils.js (hunk 2 above)
Comment (rotemee) [reply packages/elementor-ui/components/utils.js:34]: Done.

File: packages/elementor-ui/components/utils.js (hunk 1 above)
Comment (rotemee) [reply packages/elementor-ui/components/utils.js]: Done (entire new solution for caching components).

**Pull request merged**